src/main/kotlin/org/rust/lang/core/resolve/scope/RustResolveScope.kt

```diff
@@ -1,9 +1,256 @@
 package org.rust.lang.core.resolve.scope
 
-import org.rust.lang.core.psi.RustCompositeElement
-import org.rust.lang.core.psi.RustNamedElement
+import com.intellij.openapi.module.Module
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.psi.util.PsiTreeUtil
+import com.intellij.psi.util.PsiUtilCore
+import org.rust.cargo.util.AutoInjectedCrates
+import org.rust.cargo.util.cargoProject
+import org.rust.cargo.util.getPsiFor
+import org.rust.cargo.util.preludeModule
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.impl.RustFile
+import org.rust.lang.core.psi.impl.mixin.isStarImport
+import org.rust.lang.core.psi.impl.rustMod
+import org.rust.lang.core.psi.util.module
 
 interface RustResolveScope : RustCompositeElement {
-    val declarations: Collection<RustNamedElement>
+    data class Context(
```

Comment:
>Entry instead of checking whether we're listing an aliased decl.

Checking for aliases specifically feels smelly. Just look at the [assign](https://github.com/intellij-rust/intellij-rust/blob/master/src/main/kotlin/org/rust/lang/core/resolve/RustResolveEngine.kt#L314-L337) method: all this casts looks ugly to me because they try restore the information which was simply forgotten when we were processing scopes.

Not sure that visibility is derivable from PSI: you'll need to know the whole path which was used to resolve the item. 

The same thing with priority: you need to know not only the item itself, but the way which was used for importing: was it a start import or perhaps a direct import.